[Connector API] Update docs filtering, configuration, delete, scheduling (#107842)
elastic/elasticsearch: docs/reference/connector/apis/update-connector-filtering-api.asciidoc

preview::[]

> Updates the draft `filtering` configuration of a connector and marks the draft validation state as `edited`. The draft filtering can be then activated after a successful validation of the draft by the {enterprise-search-ref}/connectors.html[Elastic connector service].

Reviewer: ```suggestion
Updates the draft `filtering` configuration of a connector and marks the draft validation state as `edited`. The draft filtering can be then activated after a successful validation of the draft by the Elastic connector service.
```

Think this link might be jarring here.

Reviewer: Again "The filtering configuration can be activated once validated by the Elastic connector service." feels passive, what does the user concretely have to do to activate?

> Updates the draft `filtering` configuration of a connector and marks the draft validation state as `edited`. The draft filtering can be then activated after a successful validation of the draft by the {enterprise-search-ref}/connectors.html[Elastic connector service].

Reviewer: I wonder if we should make it explicit earlier that filtering is synonymous with sync rules

Author: ++

yeah, maybe a paragraph below we should add extra paragraph that `rules` == `basic sync rules` , `advanced_snippet` == `advanced sync rules`?

Learn more about filtering for connectors in the {enterprise-search-ref}/sync-rules.html[sync rules] documentation.

[[update-connector-filtering-api-request]]
==== {api-request-title}

`PUT _connector/<connector_id>/_filtering`
`PUT _connector/<connector_id>/_filtering/_activate`

[[update-connector-filtering-api-prereq]]
==== {api-prereq-title}

* To sync data using self-managed connectors, you need to deploy the {enterprise-search-ref}/build-connector.html[Elastic connector service] on your own infrastructure. This service runs automatically on Elastic Cloud for native connectors.
* The `connector_id` parameter should reference an existing connector.

> * Draft filtering needs to be validated by the Elastic connector service before its activation. The `draft.validation.state` needs to be `valid`.

Reviewer: "Draft filtering needs to be validated by the Elastic connector service" reads _passively_, but as a user what do I actively have to do here to ensure this validation happens?

Author: same as with configuration I replied above:

They need to have the configured service running:

- for native connectors the schema will be populated by the connector service automatically
- for self-managed connectors, you need to populate config.yml and start the service

The upon adding new draft filtering, the framework will pick it up, validate (note advanced rules could be validating for couple of mins 💀 ) and then the draft will be in `valid` state. 

then `_filtering/_activate` endpoint will work, otherwise it will throw and error saying:
```
Filtering draft needs to be validated by the connector service before activation. Current filtering draft validation state [edited] is not equal to [valid].
```

Reviewer: OK cool I mean I think you're covered with "the connector service must be running". The API calls make no sense without the connector service running anyway, so I think we can

Re. the error message I would also be explicit saying something like "Make sure connector service is running."

[[update-connector-filtering-api-request-body]]
==== {api-request-body-title}

`rules`::
(Optional, array of objects)
An array of {enterprise-search-ref}/sync-rules.html#sync-rules-basic[basic sync rules], each with the following sub-attributes:

> - `id` (Required, string) +

Reviewer: nit: we normally use * for lists in asciidoc, but if it's rendering OK not a big deal

[[update-connector-filtering-api-example]]
==== {api-examples-title}

> The following example updates the draft {enterprise-search-ref}/sync-rules.html#sync-rules-basic[basic sync rules] for the Google Drive connector with ID `my-g-drive-connector`. All Google Drive files with `.txt` extension will be skipped:

Reviewer: ```suggestion
The following example updates the draft {enterprise-search-ref}/sync-rules.html#sync-rules-basic[basic sync rules] for a Google Drive connector with ID `my-g-drive-connector`. All Google Drive files with `.txt` extension will be skipped:
```

    "result": "updated"
}
----

> The following example updates the draft advanced sync rules for a MySQL connector with id `my-sql-connector`. Advanced sync rules are specific to each connector type. Refer to the references for connectors that support {enterprise-search-ref}/sync-rules.html#sync-rules-advancedl[advanced sync rules] for syntax and examples.

Reviewer: ```suggestion
The following example updates the draft advanced sync rules for a MySQL connector with id `my-sql-connector`. Advanced sync rules are specific to each connector type. Refer to the references for connectors that support {enterprise-search-ref}/sync-rules.html#sync-rules-advanced[advanced sync rules] for syntax and examples.
```

My fault 👿

> The following example updates the draft {enterprise-search-ref}/sync-rules.html#sync-rules-advanced[advanced sync rules] for the MySQL connector with id `my-sql-connector`.  Advanced sync rules are specific to each connector type. For more details on the syntax and rules specific to the MySQL connector in our example, you can refer to the {enterprise-search-ref}/connectors-mysql.html#connectors-mysql-sync-rules[MySQL connector documentation] .

Reviewer: ```suggestion
The following example updates the draft {enterprise-search-ref}/sync-rules.html#sync-rules-advanced[advanced sync rules] for a MySQL connector with id `my-sql-connector`.  Advanced sync rules are specific to each connector type. For more details on the syntax and rules specific to the MySQL connector in our example, you can refer to the {enterprise-search-ref}/connectors-mysql.html#connectors-mysql-sync-rules[MySQL connector documentation] .
```

> The following example updates the draft {enterprise-search-ref}/sync-rules.html#sync-rules-advanced[advanced sync rules] for the MySQL connector with id `my-sql-connector`.  Advanced sync rules are specific to each connector type. For more details on the syntax and rules specific to the MySQL connector in our example, you can refer to the {enterprise-search-ref}/connectors-mysql.html#connectors-mysql-sync-rules[MySQL connector documentation] .

Reviewer: ```suggestion
The following example updates the draft advanced sync rules for a MySQL connector with id `my-sql-connector`. Advanced sync rules are specific to each connector type. Refer to the references for connectors that support {enterprise-search-ref}/sync-rules.html#sync-rules-advancedl[advanced sync rules] for syntax and examples.
```

Author: ```{enterprise-search-ref}/connectors-mysql.html#connectors-mysql-sync-rules[MySQL connector documentation] .```

Ahh I commited this but now I see that you changed this to a generic documentation of advanced sync rules - since they are connector-specific I wanted to link them to mysql definition of advances sync rule.

Reviewer: Yeah that was an opinionated choice to just say "this is an example, not all connectors support advanced sync rules, here's the top-level docs page that lists connectors that support advanced rules"

Feel free to change around again if you think it makes sense to have specific link :)

[source,console]
----
PUT _connector/my-sql-connector/_filtering
{
    "advanced_snippet": {
        "value": [{
            "tables": [
                "users",
                "orders"
            ],
            "query": "SELECT users.id AS id, orders.order_id AS order_id FROM users JOIN orders ON users.id = orders.user_id"
        }]
    }
}
----

[source,console-result]
----
{
    "result": "updated"
}
----

////
[source, console]
--------------------------------------------------
PUT _connector/my-sql-connector/_filtering/_validation
{
  "validation": {
    "state": "valid",
    "errors": []
  }
}
--------------------------------------------------
// TEST[continued]
////

Note, you can also update draft `rules` and `advanced_snippet` in a single request.

> Once the draft is updated, its validation state is set to `edited`. Connector service will then validate the rules and report the validation state as either `invalid` or `valid`. If the state is `valid`, the draft filtering can be activated with:

Reviewer: ```suggestion
Once the draft is updated, its validation state is set to `edited`. The connector service will then validate the rules and report the validation state as either `invalid` or `valid`. If the state is `valid`, the draft filtering can be activated with:
```

nitty nit
